Add integration test for hard linksPreserve hard links

uilianries opened the pull request:
Privet @SSE4 !!

I've added some changes to your work.

First, `uploader` is only the top of the iceberg, when are packaging to upload from package folder, all files there are copies already.

As we can see, I needed to update more parts, to detect hard links and preserve them. Copying from source folder to build folder and build to package are different flows and require different methods.
I also added a functional test to validate all conan create flow, it reproduces the original bug reported by the user.

There is only one case which we can't prevent for now, and maybe we should not fix:

When we pass `self.copy("*")` we are passing a list of files, so we can compare if there is a hard link on the origin, in middle of those files listed to be copied.
However, if we copy file by file, we can't compare directly, we would need to do a post package verification, like checking if a file in package folder is the same than in build folder or source folder (no_copy_sources case) and if it's a hard link on the origin. I think it's too much.

SSE4 commented on conans/test/unittests/client/util/files_test.py:
```diff
@@ -65,33 +49,3 @@ def test_path_exists(self):
         self.assertTrue(path_exists(new_path, tmp_dir))
         self.assertFalse(path_exists(os.path.join(tmp_dir, "capsdir"), tmp_dir))
-
-    @unittest.skipUnless(hardlinks_supported(), "requires hard-links")
```
why remove this test?

uilianries commented:
My mistake, I've reverted to develop branch because I added an empty to
that file. I should revert it to your branch instead.

On Mon, Nov 23, 2020, 18:22 SSE4 <notifications@github.com> wrote:

> *@SSE4* commented on this pull request.
> ------------------------------
>
> In conans/test/unittests/client/util/files_test.py
> <https://github.com/SSE4/conan/pull/12#discussion_r529004811>:
>
> > @@ -65,33 +49,3 @@ def test_path_exists(self):
>          mkdir(new_path)
>          self.assertTrue(path_exists(new_path, tmp_dir))
>          self.assertFalse(path_exists(os.path.join(tmp_dir, "capsdir"), tmp_dir))
> -
> -    @unittest.skipUnless(hardlinks_supported(), "requires hard-links")
>
> why remove this test?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SSE4/conan/pull/12#pullrequestreview-536855724>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABFFAHNCLBTXITB3DEUBUVTSRLHCTANCNFSM4T775RWA>
> .
>

SSE4 commented on conans/client/file_copier.py:
```diff
@@ -238,6 +241,16 @@ def _copy_files(files, src, dst, keep_path, symlinks):
+                        os.link(abs_src_name, abs_dst_name)
+                    except OSError as error:
+                        # e.g. OSError: [Errno 18] Invalid cross-device link
+                        raise ConanException(
```
I'd say it should fall back to the regular copy

uilianries replied:
Regular copy won't prevent hard link, if we don't change the packaging flow, we won't be able to identify hard links before compressing files.

SSE4 replied:
I am not sure I understand. I mean doing something like:
```
try:
    os.link(abs_src_name, abs_dst_name)
except OSError:
    # file system doesn't support hard-links
    shutil.copy2(abs_src_name, abs_dst_name)
```
I think cross-file-system `os.link` may fail in many cases, especially if you're copying into network drive or flash drive.

SSE4 commented on conans/util/files.py:
```diff
@@ -435,10 +435,16 @@ def link_to_rel(pointer_src):
             if os.path.islink(src_file):
                 link_to_rel(src_file)
+            elif os.stat(os.path.join(src_dir, file_)).st_ino in inodes:
+                os.link(src_file, dst_file)
```
the same, we need fallback

uilianries replied:
ditto
